lib/watir/locators/element/locator.rb

```diff
@@ -123,64 +145,44 @@ def fetch_value(element, how)
           end
         end
 
-        def all_elements
-          locate_elements(:xpath, ".//*")
-        end
-
-        def wd_find_by_regexp_selector(selector, filter)
-          query_scope = ensure_scope_context
-          rx_selector = delete_regexps_from(selector)
-
-          if rx_selector.key?(:label) && selector_builder.should_use_label_element?
-            label = label_from_text(rx_selector.delete(:label)) || return
-            if (id = label.attribute('for'))
-              selector[:id] = id
-            else
-              query_scope = label
+        def filter_elements(elements, selector, filter: :first)
+          if filter == :first
+            idx = selector.delete(:index) || 0
+            if idx.negative?
+              elements.reverse!
+              idx = idx.abs - 1
             end
-          end
-
-          how, what = selector_builder.build(selector)
-
-          unless how
-            raise Error, "internal error: unable to build Selenium selector from #{selector.inspect}"
-          end
-
-          if how == :xpath && can_convert_regexp_to_contains?
-            rx_selector.each do |key, value|
-              next if key == :tag_name || key == :text
 
-              predicates = regexp_selector_to_predicates(key, value)
-              what = "(#{what})[#{predicates.join(' and ')}]" unless predicates.empty?
-            end
+            # Lazy evaluation to avoid fetching values for elements that will be discarded
+            matches = elements.lazy.select { |el| matches_selector?(el, selector) }
+            matches.take(idx + 1).to_a[idx]
+          else
+            elements.select { |el| matches_selector?(el, selector) }
           end
-
-          elements = locate_elements(how, what, query_scope)
-          filter_elements_by_regex(elements, rx_selector, filter)
-        end
-
-        def filter_elements_by_locator(elements, visible = nil, visible_text = nil, idx = nil, tag_name: nil, filter: :first)
```

Comment:
Getting rid of these methods means I need to update `watigiri` at the very least. We'll need to ensure `watizzle` & `watir_css` also work, even though I don't think anyone uses those?